File: assets/player/states/dead.ts
const scriptId = Script.getInfo().path;

const DYING_PLAYERS = "dyingPlayers";

/** Responsible for transitioning players to the dead state whenever they are killed */
export default {
  playerStateTransition() {
    const players = world.query(PlayerState, PlayerKilled);

    // Transition all players that have been killed to this state
    for (const { entity } of players) {
      if (!Script.entityListContains(DYING_PLAYERS, entity)) {
        const [playerState] = players.get(entity);
        playerState.id = Assets.getAbsolutePath("./dead.ts");
        Script.addEntityToList(DYING_PLAYERS, entity);
      }
    }
  },
  handlePlayerState() {
    const players = world.query(PlayerState, AnimationBankSprite);

    for (const {
      entity,
      components: [playerState, animationBankSprite],
    } of players) {
      // In this state
      if (playerState.id != scriptId) continue;

      // Set animation when we enter the state
      if (playerState.age == 0) {
        animationBankSprite.current_animation = "death_1";
      }

      // Despawn player after 1.5 seconds ( 90 frames )
      if (playerState.age >= 90) {
        Script.removeEntityFromList(DYING_PLAYERS, entity);
        Player.despawn(entity);
        Script.setEntityState(entity, undefined);
      }
    }
  },
};

File: src/player/state/states/dead.rs
use crate::player::{PlayerDespawnCommand, PlayerKilled};

use super::*;

pub const ID: &str = "core:dead";

pub fn player_state_transition(mut players: Query<&mut PlayerState, With<PlayerKilled>>) {
    // Transition all killed players to this state
    for mut player_state in &mut players {
        if player_state.id != ID {
            player_state.id = ID.into();
        }
    }
}

pub fn handle_player_state(
    mut commands: Commands,
    mut players: Query<(Entity, &PlayerState, &mut AnimationBankSprite), With<PlayerKilled>>,
) {
    for (entity, state, mut sprite) in &mut players {
        if state.age == 0 {
            sprite.current_animation = "death_1".into();
        }

        if state.age >= 80 {
            commands.add(PlayerDespawnCommand::new(entity));
        }
    }
}
